### DISCUSSION TITLE
Parsing multi form data never finish

### DISCUSSION THREAD

**Discussion by nikis** [vercel/next.js#45712] [Help]
Hello, I have strange problem that I can't seem to find an answer to or debug. The tests I do is with next@13.0.1

When I upload certain files the request is pending and nothing happens. For example if I upload an empty text or small image file everything is fine but if I try to upload a video mp4 file this problem happens.

```
import { createSchema, createYoga } from 'graphql-yoga'
import { PageConfig } from 'next'

export const config: PageConfig = {
    api: {
        bodyParser: false
    }
}

export default createYoga({
    graphqlEndpoint: '/api/test',
    schema: createSchema({
        typeDefs: /* GraphQL */ `
        
        scalar File

        type Query {
            test:String
        }
        
        type Mutation {
            upload(file: File!): String!
        }
    `,
        resolvers: {
            Mutation: {
                upload: async (_, { file }: { file: File }) => {
                    console.log(file)
                    return file.name
                }
            }
        }
    })
})
```

With this curl request everything works fine and i got: {"data":{"upload":"wave.png"}}
```
curl localhost:3000/api/test \
-F operations='{ "query": "mutation ($file: File!) { upload(file: $file) }", "variables": { "file": null } }' \
-F map='{ "0": ["variables.file"] }'   -F 0=@wave.png
``` 

In this case i wait for the request to finish but it never finish
```
curl localhost:3000/api/test \
-F operations='{ "query": "mutation ($file: File!) { upload(file: $file) }", "variables": { "file": null } }' \
-F map='{ "0": ["variables.file"] }'   -F 0=@video.mp4
``` 

If i remove yoga server and export simple handler, request never stuck and everything seems to work. 
```
import { NextApiHandler, PageConfig } from 'next'

export const config: PageConfig = {
    api: {
        bodyParser: false
    }
}

const handler:NextApiHandler = (req, res) => {
    res.send(req.headers['content-type'])
}

export default handler
```

I tried with other libraries like busboy, multer, formidable and found the same behavior, parse callbacks is never called and request never finish.

I would greatly appreciate it if someone could help me.

**Comment by icyJoseph** [comment]
I think the problem is that the handler cannot tell that the work has been resolved. I remember running into a similar issue with Formidable, in the end I just made my own formidable wrapper, and I made sure to return a response from there. 

Could you try with something like this? 

```ts
import {Writable} from 'stream';

import formidable from 'formidable';
import {NextApiRequest, NextApiResponse, PageConfig} from 'next';

const formidableConfig = {
    keepExtensions: true,
    maxFileSize: 10_000_000,
    maxFieldsSize: 10_000_000,
    maxFields: 2,
    allowEmptyFiles: false,
    multiples: false,
};

// promisify formidable
function formidablePromise(
    req: NextApiRequest,
    opts?: Parameters<typeof formidable>[0]
): Promise<{fields: formidable.Fields; files: formidable.Files}> {
    return new Promise((accept, reject) => {
        const form = formidable(opts);

        form.parse(req, (err, fields, files) => {
            if (err) {
                return reject(err);
            }
            return accept({fields, files});
        });
    });
}

const fileConsumer = <T = unknown>(acc: T[]) => {
    const writable = new Writable({
        write: (chunk, _enc, next) => {
            acc.push(chunk);
            next();
        },
    });

    return writable;
};

// inside the handler

async function handler(req:NextApiRequest, res: NextApiResponse) {
    try {
        const chunks: never[] = [];

        const {fields, files} = await formidablePromise(req, {
            ...formidableConfig,
            // consume this, otherwise formidable tries to save the file to disk
            fileWriteStreamHandler: () => fileConsumer(chunks),
        });

       // do something with the files
       const contents = Buffer.concat(chunks); 

     return res.status(204).end()
     } catch(e) {
      // handle errors
     }
}

// and don't forget
export const config: PageConfig = {
    api: {
        bodyParser: false,
    },
};
```

Do you see this issue in development mode? I wonder if the problem has to with the size of the file? Is the video very large? Maybe the parsing errors out but it never rejects and the handler just hangs waiting for a resolve/reject?

**Reply by nikis** [reply]
Its happen in development and production. Also its happening with random files , but in general I have 2-3 files with which it goes and with the others there is this problem.

I try with the above example, but still same problem.

**Reply by icyJoseph** [reply]
Wow, do you think you can try with an express server or something like that and try to upload those files to it? This is rather strange. Does it happen still if you upgrade the version? Do you have middleware?

**Reply by nikis** [reply]
Yes, i try with node http server and still the same problem. I just post issue here: https://github.com/node-formidable/formidable/issues/923
